File: .local/bin/pwcalculator.lua
#!/bin/lua

-- local inspect = require('pl.import_into')().pretty.write
local regex = require('regex')
local GLib = require("lgi").GLib

local secret = arg[1]
local domain = arg[2]

local function hex2bytearray(hex)
    local re = regex.new(".{2}", "g")
    local arr, err = re:matches(hex)
    if err then error(err) end

    local bytes = {}

    for i, j in pairs(arr) do
        table.insert(bytes, (tonumber(j, 16) or 0))
        if #bytes == 20 then break end
    end

    while #bytes < 20 do table.insert(bytes, 0) end

    return bytes
end

local function sha1bytes(text)
    local c = GLib.Checksum.new(GLib.ChecksumType.SHA1)
    c:update(text)
    return hex2bytearray(c:get_string())
end

local text = (secret or " ") .. (domain or " ")
local sha1 = sha1bytes(text)
local base64 = GLib.base64_encode(sha1);
io.write(string.sub(base64, 1, 16))

File: .local/bin/pwcalculator.js
#!/bin/gjs

const GLib = imports.gi.GLib;

function hex2bytearray(hex) {
  var array = hex.match(/.{2}/g);
  var bytes = new Uint8Array(20);
  for (var i = 0; i < array.length; i++) bytes[i] = parseInt(array[i], 16);
  return bytes;
}

function sha1bytes(text) {
  var c = new GLib.Checksum(GLib.ChecksumType.SHA1);
  c.update(text);
  //return c.get_digest();
  return hex2bytearray(c.get_string());
}

function sha1(secret, domain, length) {
  if (secret == "" || domain == "") return "";
  var text = secret + domain;
  var sha1 = sha1bytes(text);
  var base64 = GLib.base64_encode(sha1);
  return base64.substring(0, length);
}

// if (ARGV.length === 0) {
//   print("No arguments provided.");
// } else {
//   // Iterate through each command-line argument
//   for (let i = 0; i < ARGV.length; i++) {
//     let arg = ARGV[i];
//     print("Argument " + i + ": '" + arg + "'");
//   }
// }

// print(ARGV[-1]);
//
// print(sha1bytes(ARGV[1]));
// print(sha1bytes(ARGV[2]));

print(sha1(ARGV[0], ARGV[1], 16));
